Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -153,567 +131,429 @@ final class ConnectionPool {
-                    self.stateLock.withLock {
-                        guard connection.mustRunDefaultCloseCallback else {
-                            return nil
-                        }
-                        switch result {
-                        case .success:
-                            return self.state.removeClosedConnection(connection)
+    var closePromise: EventLoopPromise<Void>
+
+    var state: ConnectionsState
+
+    /// Creates a new `HTTP1ConnectionProvider`
+    ///
+    /// - parameters:
+    ///     - key: The `Key` (host, scheme, port) this provider is associated to
+    ///     - configuration: The client configuration used globally by all requests
+    ///     - initialConnection: The initial connection the pool initializes this provider with
+    ///     - pool: The pool this provider belongs to
+    init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, pool: ConnectionPool) {
+        self.eventLoop = eventLoop
+        self.configuration = configuration
+        self.key = key
+        self.pool = pool
+        self.closePromise = eventLoop.makePromise()
+        self.state = .init(eventLoop: eventLoop)
+    }
 
+    deinit {
+        self.state.assertInvariants()
+    }
+
+    private func execute(_ action: Action) {
+        switch action {
+        case .lease(let connection, let waiter):
+            // if connection is became inactive, we create a new one.
+            connection.cancelIdleTimeout().whenComplete { _ in
+                if connection.isActiveEstimation {
+                    waiter.promise.succeed(connection)
+                } else {
+                    self.makeChannel(preference: waiter.preference).whenComplete { result in
+                        switch result {
+                        case .success(let channel):
+                            self.connect(using: channel, waiter: waiter, replacing: connection)
                         case .failure(let error):
-                            preconditionFailure("Connection close future failed with error: \(error)")
+                            self.connectFailed(error: error, waiter: waiter)
                         }
                     }
                 }
-
-                if let action = action {
-                    switch action {
-                    case .makeConnectionAndComplete(let el, let promise):
-                        self.makeConnection(on: el).cascade(to: promise)
-                    case .none:
-                        break
+            }
+        case .create(let waiter):
+            self.makeChannel(preference: waiter.preference).whenComplete { result in
+                switch result {
+                case .success(let channel):
+                    self.connect(using: channel, waiter: waiter)
+                case .failure(let error):
+                    self.connectFailed(error: error, waiter: waiter)
+                }
+            }
```

Comment:
This block is repeated a few times, it seems like we could factor it out a bit.

Reply:
Good idea, done, thanks!